File: mayonez-base/src/main/java/mayonez/graphics/renderer/GLRenderer.java
package mayonez.graphics.renderer;

import mayonez.Preferences;
import mayonez.annotations.EngineType;
import mayonez.annotations.UsesEngine;
import mayonez.graphics.GLCamera;
import mayonez.graphics.GLRenderable;
import mayonez.graphics.RenderBatch;
import mayonez.io.Assets;
import mayonez.io.image.Shader;

import java.awt.*;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

/**
 * Contains methods for uploading sprite and shape data to the GPU for renderers using OpenGL.
 */
@UsesEngine(EngineType.GL)
public abstract class GLRenderer implements Renderer {

    // GPU Resources
    private final List<RenderBatch> batches;
    protected final Shader shader;
    protected final int[] textureSlots; // support multiple texture IDs in batch

    public GLRenderer(String shaderFile) {
        this.batches = new ArrayList<>();
        this.shader = Assets.getShader(shaderFile);
        textureSlots = new int[Preferences.getMaxTextureSlots()];
        for (var i = 0; i < this.textureSlots.length; i++) this.textureSlots[i] = i; // ints 0-7
    }

    // Renderer Methods

    @Override
    public void start() {
        batches.clear();
    }

    @Override
    public final void render(Graphics2D g2) {
        preRender();
        rebuffer();
        batches.forEach(RenderBatch::render); // Draw everything
        postRender();
    }

    /**
     * Upload resources to the GPU.
     */
    protected void preRender() {
        // Bind shader and upload uniforms
        shader.bind();
        var camera = (GLCamera) getCamera();
        shader.uploadMat4("uProjection", camera.getProjectionMatrix());
        shader.uploadMat4("uView", camera.getViewMatrix());
    }

    /**
     * Sort all image data into batches.
     */
    protected final void rebuffer() {
        batches.forEach(RenderBatch::clear); // Prepare batches
        createBatches();
        batches.forEach(RenderBatch::upload); // Finalize batches
        batches.sort(Comparator.comparingInt(RenderBatch::getZIndex)); // Sort batches by z-index
    }

    /**
     * Free resources from the GPU.
     */
    protected void postRender() {
        shader.unbind(); // Unbind everything
    }

    @Override
    public void stop() {
        batches.forEach(RenderBatch::delete);
    }

    // Batch Helper Methods

    /**
     * Sort image data into render batches.
     */
    protected abstract void createBatches();

    protected final RenderBatch getAvailableBatch(GLRenderable object) {
        for (RenderBatch batch : batches) {
            if (doesBatchFit(batch, object)) return batch;
        }
        // If all batches full
        return createNewBatch(object);
    }

    private boolean doesBatchFit(RenderBatch batch, GLRenderable object) {
        var fitsVertices = batch.hasRoom();
        var samePrimitive = batch.getPrimitive() == object.getPrimitive();
        var sameZIndex = batch.getZIndex() == object.getZIndex();
        var fitsTexture = batch.hasTexture(object.getTexture()) || batch.hasTextureRoom();
        return fitsVertices && samePrimitive && sameZIndex && fitsTexture;
    }

    private RenderBatch createNewBatch(GLRenderable object) {
        RenderBatch batch = new RenderBatch(object.getBatchSize(), object.getZIndex(), object.getPrimitive());
        batch.clear();
        batches.add(batch);
        return batch;
    }

}

File: mayonez-base/src/main/kotlin/mayonez/graphics/renderer/GLRenderer.kt
package mayonez.graphics.renderer

import mayonez.assets.*
import mayonez.graphics.*
import mayonez.graphics.batch.*
import mayonez.graphics.shader.*
import java.awt.*

/**
 * A base renderer for OpenGL that uploads sprite and shape data to the
 * GPU.
 *
 * @author SlavSquatSuperstar
 */
@UsesEngine(EngineType.GL)
abstract class GLRenderer(shaderFile: String) : Renderer {

    // GPU Resources
    private val batches: MutableList<RenderBatch> = ArrayList()
    protected val shader: Shader = Assets.getAsset(shaderFile, Shader::class.java)!!

    protected val textureSlots: IntArray = IntArray(RenderBatch.MAX_TEXTURE_SLOTS) { it }
    // support multiple texture IDs in batch ints 0-7

    // Renderer Methods
    override fun start() {
        batches.clear()
    }

    override fun render(g2: Graphics2D?) {
        preRender()
        rebuffer()
        batches.forEach(RenderBatch::drawBatch)
        postRender()
    }

    /** Clear the screen and upload resources to the GPU. */
    protected open fun preRender() {
        shader.bind()
    }

    /** Sort all image data into batches. */
    protected open fun rebuffer() {
        batches.forEach(RenderBatch::clearVertices) // Prepare batches
        createBatches()
        batches.forEach(RenderBatch::uploadVertices) // Finalize batches
        batches.sortBy(RenderBatch::getZIndex) // Sort batches by z-index
    }

    /** Finish drawing and free resources from the GPU. */
    protected open fun postRender() {
        shader.unbind() // Unbind everything
    }

    override fun stop() {
        batches.forEach(RenderBatch::deleteBatch)
    }

    // Batch Helper Methods

    /** Sort image data into render batches. */
    protected abstract fun createBatches()

    protected fun GLRenderable.getAvailableBatch(): RenderBatch {
        return batches.find { this.fitsInBatch(it) }
            ?: this.createNewBatch() // If all batches full
    }

    private fun GLRenderable.fitsInBatch(batch: RenderBatch): Boolean {
        return batch.hasVertexRoom() && (batch.primitive == this.primitive)
                && (batch.zIndex == this.zIndex) && batch.canFitTexture(this)
    }

    private fun RenderBatch.canFitTexture(renderable: GLRenderable): Boolean {
        return this.hasTexture(renderable.texture) || this.hasTextureRoom()
    }

    private fun GLRenderable.createNewBatch(): RenderBatch {
        val batch = RenderBatch(batchSize, zIndex, primitive)
        batches.add(batch)
        return batch
    }

}